### PR TITLE
Add collision avoidance nullspace component to IK solver

### PR DESCRIPTION
**Pull request by sea-bass**
This PR adds a collision avoidance component to the differential IK solver.

It does this by finding a collision Jacobian and pushing things away from that direction. Here's a picture to try explain.

image

I have no idea if I'm doing this right... math is hard.

### CODE REVIEW

**Review by eholum — COMMENTED** [review]
Some clarifying questions because I'm catching up. I think we need Cartesian servoing to see this in action.

File: src/pyroboplan/ik/nullspace_components.py
Comment (eholum) [inline src/pyroboplan/ik/nullspace_components.py:129]: Should this also check if the distance is 0 (ie, in collision)? Otherwise I think things break down?

**Comment by eholum** [comment]
> I think we need Cartesian servoing to see this in action.

This is just amazing.

https://github.com/sea-bass/pyroboplan/assets/7966037/b57de32e-15e6-498f-aa0a-c912dd3e9b8b

https://github.com/sea-bass/pyroboplan/assets/7966037/0e83aaa2-2f9f-4ee3-bc93-b4efa19565db

File: src/pyroboplan/ik/nullspace_components.py
Comment (sea-bass) [reply src/pyroboplan/ik/nullspace_components.py:129]: I need to check again, but I thought the distances were negative if the objects are overlapping

File: src/pyroboplan/core/utils.py
Comment (sea-bass) [reply src/pyroboplan/core/utils.py:35]: I think it's just more efficient to reuse existing structs than create new ones each collision check...

But as I understand, the motivation is for example to have 1 model and multiple data structs in use cases like multithreading

File: src/pyroboplan/ik/differential_ik.py
Comment (sea-bass) [reply src/pyroboplan/ik/differential_ik.py:61]: Yeah, hence the TODO to use actual joint groups

Comment (eholum) [reply src/pyroboplan/ik/differential_ik.py:61]: I never said I could read!

File: src/pyroboplan/ik/nullspace_components.py
Comment (sea-bass) [reply src/pyroboplan/ik/nullspace_components.py:165]: No issue with delta_q on the grippers as those joints don't affect the pose of other frames.

The one line I am not clear about, and is likely the source of error, is that `toActionMatrix()` one. What we need to do is transform a Jacobian from one frame to the other and I am not certain that's the way to do it.

The other lines after that are just converting Cartesian velocity to joint velocity like everywhere else.

Comment (sea-bass) [reply src/pyroboplan/ik/nullspace_components.py]: I put the references in, but this was totally wrong. See

https://typeset.io/pdf/a-collision-free-mpc-for-whole-body-dynamic-locomotion-and-1l6itpfk.pdf

https://laas.hal.science/hal-04425002

And for the second paper, there is actually a C++ implementation here: https://github.com/agimus-project/colmpc/blob/main/include/colmpc/residual-distance-collision.hxx

File: src/pyroboplan/ik/nullspace_components.py
Comment (sea-bass) [reply src/pyroboplan/ik/nullspace_components.py]: I checked again. They are not. So now I am actually doing a collision check as well and getting the normal/penetration depth from there.

It's unclear to me if I did it right, but it's a step in the right direction...

File: src/pyroboplan/ik/differential_ik.py
Comment (sea-bass) [reply src/pyroboplan/ik/differential_ik.py:254]: I'm sure it does, but it will probably be a bit more manual effort and probably takes away from the pedagogical goals of this library... which are already being severely violated with this PR haha

File: src/pyroboplan/ik/nullspace_components.py
Comment (sea-bass) [reply src/pyroboplan/ik/nullspace_components.py]: This is now gone woo

**Comment by sea-bass** [comment]
> > I think we need Cartesian servoing to see this in action.
> 

Curious if the latest changes make that servoing better. The math seems more sound now.

When I carve out my next time block to work on this, I'm going to try validate the HPP-FCL collision/distance queries to see if they are correct.

**Comment by sea-bass** [comment]
I just found something that might be necessary to iron out to do this correctly:
https://github.com/humanoid-path-planner/hpp-fcl/issues/587

I think relying on just the collision result normal/penetration depth may not have sufficient information to get the direction of the collision vector correct.

**Review by eholum — COMMENTED** [review]
This is so interesting. Obviously my stupid, stupid "servo" needs some work, but it's really nifty to see how this pushes around obstacles. There isn't a joint jump here, I just don't have velocity limits,


https://github.com/sea-bass/pyroboplan/assets/7966037/424843d9-1f8f-4c99-9963-c1f29f1961ff

**Comment by sea-bass** [comment]
Yeah, so the way these tasks are stacked is that the IK takes priority over the collision checking.

Meaning that regardless of collision, the IK solver will reduce its error (per first derivatives). BUT if there is redundancy in the degrees of freedom, it will null out any motion that is towards collision and use the other DOFs instead.

If this were to be cast an actual collision-avoiding servo, I think the tasks would be flipped over so that you first do the collision avoidance, and use the nullspace of that task to actually track your target commands.

File: src/pyroboplan/core/utils.py
Comment (eholum) [inline src/pyroboplan/core/utils.py:426]: super super nit: could be a helper function to compute the jacobian at a point to reduce code duplication

File: src/pyroboplan/core/utils.py
Comment (eholum) [inline src/pyroboplan/core/utils.py:386]: Will this raise an exception if they're missing?

File: src/pyroboplan/ik/nullspace_components.py
Comment (eholum) [inline src/pyroboplan/ik/nullspace_components.py:94]: 👍👍👍

File: src/pyroboplan/ik/nullspace_components.py
Comment (eholum) [inline src/pyroboplan/ik/nullspace_components.py:141]: `np.finfo(float).eps`?

**Review by eholum — DISMISSED** [review]
I am not claiming that this is bug free - but I've seen it in action and I _think_ I understand the mechanics of it. And I learned something new! I think others will as well.

I added some nit comments, nothing big. I will defer to your expert opinion to press that merge button.

File: src/pyroboplan/core/utils.py
Comment (sea-bass) [reply src/pyroboplan/core/utils.py:386]: nope. This is sadly just how Pinocchio works, that the model and collision data needs to be initialized or else it has all default values.

File: src/pyroboplan/ik/nullspace_components.py
Comment (sea-bass) [reply src/pyroboplan/ik/nullspace_components.py:141]: I don't want to use `eps` necessarily, but I think at least making a constant is appropriate.

File: src/pyroboplan/core/utils.py
Comment (sea-bass) [reply src/pyroboplan/core/utils.py:426]: I considered this, but this is one of those situations where it makes things even more confusing than it already is, so I chose to leave it there.

**Pull request merged**
